weave-js/src/components/PagePanelComponents/Home/Browse3/datasets/DatasetVersionPage.tsx

```diff
@@ -53,12 +90,204 @@ export const DatasetVersionPage: React.FC<{
+        updates.push({pop: {index: rowIndex}});
+        updates.push({
+          insert: {
+            index: rowIndex,
+            row: cleanRow(editedRow),
+          },
+        });
+      }
+    });
+
+    deletedRows
+      .sort((a, b) => b - a)
+      .forEach(rowIndex => {
+        updates.push({pop: {index: rowIndex}});
+      });
+
+    Array.from(addedRows.values())
+      .reverse()
+      .forEach(row => {
+        updates.push({
+          insert: {
+            index: 0,
+            row: cleanRow(row),
+          },
+        });
+      });
+
+    return updates;
+  }, [editedRows, deletedRows, addedRows]);
+
+  const tableUpdate = useTableUpdate();
+  const objCreate = useObjCreate();
+
+  const handlePublish = useCallback(async () => {
+    setIsEditing(false);
+    setPublishAnchorEl(null);
+
+    const tableUpdateSpecs = convertEditsToTableUpdateSpec();
+    const tableUpdateResp = await tableUpdate(
+      projectId,
+      originalTableDigest,
+      tableUpdateSpecs
+    );
+    const tableRef = `weave:///${projectId}/table/${tableUpdateResp.digest}`;
+
+    const newObjVersion = await objCreate(projectId, objectName, {
+      ...objectVersion.val,
+      rows: tableRef,
+    });
+
+    const url = router.objectVersionUIUrl(
+      entityName,
+      projectName,
+      objectName,
+      newObjVersion,
+      undefined,
+      refExtra
+    );
+
+    toast(
+      <div style={{display: 'flex', alignItems: 'center', gap: '8px'}}>
+        <Icon name="checkmark" width={20} height={20} />
+        Published{' '}
+        <Link to={url} style={PUBLISHED_LINK_STYLES}>
+          {objectName}:v{objectVersionCount}
+        </Link>
+      </div>
+    );
+    reset();
+  }, [
+    reset,
+    refExtra,
+    router,
+    objectName,
+    objectVersionCount,
+    objectVersion.val,
+    convertEditsToTableUpdateSpec,
+    projectId,
+    objCreate,
+    tableUpdate,
+    originalTableDigest,
+    entityName,
+    projectName,
+  ]);
+
+  const renderEditingControls = () => (
+    <div className="flex gap-8">
+      <Typography
+        variant="body2"
+        sx={{
+          color: 'text.secondary',
+          display: 'flex',
+          alignItems: 'center',
+          fontSize: '14px',
+          fontFamily: 'Source Sans Pro',
+        }}>
+        Editing dataset
+        <Icon name="pencil-edit" width={14} height={14} />
+      </Typography>
+      <Button
+        title="Cancel"
+        variant="ghost"
+        size="medium"
+        icon="close"
+        onClick={handleCancelClick}>
+        Cancel
+      </Button>
+      <Button
+        title="Publish"
+        size="medium"
+        variant="primary"
+        icon="checkmark"
+        onClick={handlePublishClick}
+        disabled={
+          editedCellsMap.size === 0 &&
+          deletedRows.length === 0 &&
+          addedRows.size === 0
+        }>
+        Publish
+      </Button>
+    </div>
+  );
+
+  const renderPublishPopover = () => (
```

Comment:
this component looks a touch awkward, maybe left-bias the buttons? or make it wider? idk

<img width="328" alt="Screenshot 2025-01-09 at 10 01 31 AM" src="https://github.com/user-attachments/assets/2d318f2d-b626-4774-a5f7-a4054f070046" />

Reply:
Also the double "publish" looks a bit much? again idk

<img width="319" alt="Screenshot 2025-01-09 at 10 03 02 AM" src="https://github.com/user-attachments/assets/64ff3823-344d-4383-8c39-48f6b9369a98" />

Reply:
I put confirm instead of publish, for example